pkg/state/infra.go

```diff
@@ -116,24 +195,319 @@ func (i *infraClient) SyncConnectors(routers []string, brokers []string) ([]Conn
+	ctx := context.Background()
+
+	// Skip endpoints that are not active or do not have hosts defined
+	if endpoint.Status.Phase != v1beta2.MessagingEndpointActive || endpoint.Status.Host == "" {
+		return nil
+	}
+
+	tenantId := endpoint.Status.Host
+
+	// Configure all routers
+	err := i.applyRouters(ctx, func(router *RouterState) error {
+		addressName := addressName(tenantId, address)
+		if address.Spec.Anycast != nil {
+			err := router.EnsureAddress(&RouterAddress{
+				Name:         addressName,
+				Prefix:       fmt.Sprintf("%s/%s", tenantId, address.GetAddress()),
+				Distribution: "balanced",
+				Waypoint:     false,
+			})
+			if err != nil {
+				return err
+			}
+		} else if address.Spec.Multicast != nil {
+			err := router.EnsureAddress(&RouterAddress{
+				Name:         addressName,
+				Prefix:       fmt.Sprintf("%s/%s", tenantId, address.GetAddress()),
+				Distribution: "multicast",
+				Waypoint:     false,
+			})
+			if err != nil {
+				return err
+			}
+		} else if address.Spec.Queue != nil {
+			err := router.EnsureAddress(&RouterAddress{
+				Name:         addressName,
+				Prefix:       fmt.Sprintf("%s/%s", tenantId, address.GetAddress()),
+				Distribution: "balanced",
+				Waypoint:     true,
+			})
+			if err != nil {
+				return err
+			}
+
+			for _, broker := range address.Status.Brokers {
+				brokerState := i.brokers[broker.Host]
+				if brokerState == nil {
+					return fmt.Errorf("Unable to configure address autoLink (tenant %s address %s) for unknown broker %s", tenantId, address.GetAddress(), broker.Host)
+				}
+				err = router.EnsureAutoLink(&RouterAutoLink{
+					Name:            autoLinkName(tenantId, address, broker.Host, "in"),
+					Address:         fmt.Sprintf("%s/%s", tenantId, address.GetAddress()),
+					Direction:       "in",
+					Connection:      connectorName(brokerState),
+					ExternalAddress: fmt.Sprintf("%s/%s", tenantId, address.GetAddress()),
+				})
+				if err != nil {
+					return err
+				}
+
+				err = router.EnsureAutoLink(&RouterAutoLink{
+					Name:            autoLinkName(tenantId, address, broker.Host, "out"),
+					Address:         fmt.Sprintf("%s/%s", tenantId, address.GetAddress()),
+					Direction:       "out",
+					Connection:      connectorName(brokerState),
+					ExternalAddress: fmt.Sprintf("%s/%s", tenantId, address.GetAddress()),
+				})
+				if err != nil {
+					return err
+				}
+			}
+		} else if address.Spec.Topic != nil {
+			// TODO
+			return fmt.Errorf("Unsupported address type 'topic'")
+		} else if address.Spec.Subscription != nil {
+			// TODO
+			return fmt.Errorf("Unsupported address type 'subscription'")
+		}
+		return nil
+	})
+	if err != nil {
+		return err
+	}
+
+	err = i.applyBrokers(ctx, func(broker *BrokerState) error {
+		if address.Spec.Queue != nil {
+			// Configure relevant brokers
+			for _, brokerInfo := range address.Status.Brokers {
+				if brokerInfo.Host == broker.Host {
+					err := broker.EnsureQueue(tenantId, address)
```

Comment:
We are serialising queue creation, which I think is okay for now.  Later I think we'll need to be prepare to dispatch n createQueues and then await them in order to efficiently handle the case where there are many thousands of queues.

Reply:
Thanks, I'll add a comment about that. I think we should do some perf tests to test this bottle neck before spending much time to fix it though.

In this case, the BrokerState maintains a map of the current queues, so the management command to create the queue will only happen if the internal brokerState.queues map does not have an entry for the queue. This means that the management request will only fire on initial address creation or perhaps when migrating queues rather, as opposed to the fixed interval management queries that occur in standard-controller/agent today.